### PR TITLE
Produce resolved Ansible snippets early in the build process.

### PR DESCRIPTION
**Pull request by matejak**
This PR changes the Ansible remediations generation process. It aims to make build more transparent and reusable by exposing almost-ready remediations as individual files in the build directory. Next, it gets rid of `@ANSIBLE_TAGS@` and other placeholders and slightly reduces the amount of XSLT transforms performed during the build.

Past state:
* Ansible remediations were aggregated to the huge `fixes-ansible.xml` file as templates - they have contained those `@ANSIBLE_TAGS@` and other placeholders, they weren't valid YAML snippets.
* Later in the build process, an XSLT transform expanded those snippets to valid Ansible/YAML.

New state:
* The `@ANSIBLE_TAGS@` and some (not all) other placeholders were removed, as they are not meant to be used.
* Individual snippets are exported to `build/<product>/fixes/ansible/<rule-id>.yml` in form of valid YAML files, with `tags` and `when` populated correctly.

Regressions:
At this stage, no regressions are expected. As `pyyaml` is used in part of the processing, the indentation, quotation and interpretation of boolean values changed, so it is not trivial to compare new and old remediations. I have used profile roles with leading whitespace trimmed, which allows for quite nice comparison.

How it has been done:
* The remediation build has been slightly refactored by introducing classes - now e.g. Bash and Ansible remediations are quite different, and method override is a convenient way how to introduce additional steps to remediation build.
* The `AnsibleRemediation` class in `ssg.ansible` reads the remediation stub, looks up the resolved rule in the build directory, and adds metadata to the remediation.
* This happens during the remediation collection phase, i.e. before embedding remediations in huge XML files.
* Some XSLT is still performed later in the build, but the scope of it is now very limited.

The ugly:
There is now an `AnsibleRemediation` class in `ssg.ansible` and `ssg.build_remediations`. Although there is no code duplication, those two classes should be merged.
See also #3830.

### CODE REVIEW

File: ssg/build_remediations.py [hunk 1]
```diff
@@ -218,26 +220,94 @@ def parse_from_file_without_jinja(file_path):
...
+            self.resolved_rules, self.fix_name + ".yml")
+        if not os.path.isfile(rule_path):
+            msg = ("Ansible snippet for rule {rule_id} "
+                   "doesn't have respective rule YML at {rule_fname}"
+                   .format(rule_id=os.path.splitext(self.fix_name)[0], rule_fname=rule_path))
+            print(msg, file=sys.stderr)
+            return result
+
+        import ssg.ansible
+        from ssg import build_yaml
```
Comment (jan-cerny) [inline ssg/build_remediations.py]: Can the imports be at the beginning of file?

File: ssg/ansible.py [hunk 2]
```diff
@@ -50,3 +56,137 @@ def remove_trailing_whitespace(ansible_src):
...
+        result += _strings_to_list(new_contents)
+    if not result:
+        result = ""
+    if len(result) == 1:
+        result = result[0]
+    return result
+
+
+class AnsibleRemediation(object):
+
```
Comment (jan-cerny) [inline ssg/ansible.py]: Do you think that this script cans also replace the `@RULE_TITLE@` macro?

File: tests/unit/ssg-module/data/file_owner_grub2_cfg.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# File processed by the build system, i.e. retrieved from the build/rhel7/rules/file_owner_grub2_cfg.yml
+#
+description: 'The file <tt>/boot/grub2/grub.cfg</tt> should
+
+  be owned by the <tt>root</tt> user to prevent destruction
+
```
Comment (jan-cerny) [inline tests/unit/ssg-module/data/file_owner_grub2_cfg.yml]: Why are these empty lines there?

Comment (matejak) [reply tests/unit/ssg-module/data/file_owner_grub2_cfg.yml]: It has been taken straight from the build, so this corresponds to the produced content.

File: ssg/ansible.py (hunk 2 above)
Comment (matejak) [reply ssg/ansible.py]: The proper solution to this is replacing those with jinja macros. As it is easy, but not trivial, I propose to consider it out of scope of this PR.

File: ssg/build_remediations.py (hunk 1 above)
Comment (matejak) [reply ssg/build_remediations.py]: Unfortunately not, there is an import loop, and late imports are the only way how to make it work. The proper solution would be to unify the two remediation classes and straightening the dependency graph as a consequence.

**Comment by matejak** [comment]
retest this please

**Comment by matejak** [comment]
test this please

**Comment by matejak** [comment]
test this please

**Comment by matejak** [comment]
A reference note - extra care is needed when dealing with build steps that produce more than one outputs that is used by multiple build targets - it is advisable to mention only the latest output that is produced, and [according to cmake docs](https://cmake.org/cmake/help/v3.0/command/add_custom_command.html#add-custom-command) specify the "internal" target as dependency (as opposed to the actual produced file).

**Comment by jan-cerny** [comment]
@matejak When I left the office yesterday I had the same idea near the fake lake :smiley: :smiley:

**Comment by jan-cerny** [comment]
@matejak Great! I like your PR very much, it's very helpful. The generated content in `fixes` directories looks fine.

File: build-scripts/combine_remediations.py [hunk 4]
```diff
@@ -28,6 +28,10 @@ def parse_args():
         help="YAML file with information about the product we are building. "
         "e.g.: ~/scap-security-guide/rhel7/product.yml"
     )
+    p.add_argument(
+        "--resolved-rules", required=True,
```
[the remark this answers is not in the document]
Comment (matejak) [reply build-scripts/combine_remediations.py]: Tracked in #4027.

**Pull request merged**